src/libraries/Microsoft.Extensions.Options/gen/Parser.cs

```diff
@@ -0,0 +1,663 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+
+namespace Microsoft.Extensions.Options.Generators
+{
+    /// <summary>
+    /// Holds an internal parser class that extracts necessary information for generating IValidateOptions.
+    /// </summary>
+    internal sealed class Parser
+    {
+        private const int NumValidationMethodArgs = 2;
+
+        private readonly CancellationToken _cancellationToken;
+        private readonly Compilation _compilation;
+        private readonly Action<Diagnostic> _reportDiagnostic;
+        private readonly SymbolHolder _symbolHolder;
+        private readonly Dictionary<ITypeSymbol, ValidatorType> _synthesizedValidators = new(SymbolEqualityComparer.Default);
+        private readonly HashSet<ITypeSymbol> _visitedModelTypes = new(SymbolEqualityComparer.Default);
+
+        public Parser(
+            Compilation compilation,
+            Action<Diagnostic> reportDiagnostic,
+            SymbolHolder symbolHolder,
+            CancellationToken cancellationToken)
+        {
+            _compilation = compilation;
+            _cancellationToken = cancellationToken;
+            _reportDiagnostic = reportDiagnostic;
+            _symbolHolder = symbolHolder;
+        }
+
+        public IReadOnlyList<ValidatorType> GetValidatorTypes(IEnumerable<TypeDeclarationSyntax> classes)
+        {
+            var results = new List<ValidatorType>();
+
+            foreach (var group in classes.GroupBy(x => x.SyntaxTree))
+            {
+                SemanticModel? sm = null;
+                foreach (var typeDec in group)
+                {
+                    _cancellationToken.ThrowIfCancellationRequested();
+                    sm ??= _compilation.GetSemanticModel(typeDec.SyntaxTree);
```

Comment:
Calling `GetSemanticModel` is generally speaking expensive. The `ForAttributeWithMetadataName` method gives you the `SemanticModel` for every syntax node you want to visit so this call can be avoided altogether:

https://github.com/dotnet/runtime/blob/36439c510b779103a4a8066359d0d63dc003eed3/src/libraries/System.Text.Json/gen/JsonSourceGenerator.Roslyn4.0.cs#L40

Reply:
I have addressed this in the commit [`8ea301d` (#87587)](https://github.com/dotnet/runtime/pull/87587/commits/8ea301d5d2ffe0c927b6349c707bb5350cc3e151)